class-jquery-migrate-helper.php

```diff
@@ -0,0 +1,177 @@
+<?php
+/**
+ * Class jQuery_Migrate_Helper,
+ *
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	die( 'Invalid request.' );
+}
+
+class jQuery_Migrate_Helper {
+
+	private function __construct() {}
+
+	public static function init_actions() {
+		// To be able to replace the src, scripts should not be concatenated.
+		if ( ! defined( 'CONCATENATE_SCRIPTS' ) ) {
+			define( 'CONCATENATE_SCRIPTS', false );
+		}
+
+		$GLOBALS['concatenate_scripts'] = false;
+
+		add_action( 'wp_default_scripts', array( __CLASS__, 'replace_scripts' ), -1 );
+
+		// We need our own script for displaying warnings to run as late as possible.
+		// Print it separately after the footer scripts.
+		add_action( 'admin_enqueue_scripts', array( __CLASS__, 'register_scripts' ) );
+		add_action( 'admin_print_footer_scripts', array( __CLASS__, 'print_scripts' ), 100 );
+
+		add_action( 'admin_notices', array( __CLASS__, 'admin_notices' ) );
+		add_action( 'wp_ajax_jquery-migrate-dismiss-notice', array( __CLASS__, 'admin_notices_dismiss' ) );
+	}
+
+	// Pre-register scripts on 'wp_default_scripts' action, they won't be overwritten by $wp_scripts->add().
+	private static function set_script( $scripts, $handle, $src, $deps = array(), $ver = false, $in_footer = false ) {
+		$script = $scripts->query( $handle, 'registered' );
+
+		if ( $script ) {
+			// If already added
+			$script->src  = $src;
+			$script->deps = $deps;
+			$script->ver  = $ver;
+			$script->args = $in_footer;
+
+			unset( $script->extra['group'] );
+
+			if ( $in_footer ) {
+				$script->add_data( 'group', 1 );
+			}
+		} else {
+			// Add the script
+			if ( $in_footer ) {
+				$scripts->add( $handle, $src, $deps, $ver, 1 );
+			} else {
+				$scripts->add( $handle, $src, $deps, $ver );
+			}
+		}
+	}
+
+	/*
+	 * Enqueue jQuery migrate, and force it to be the development version.
+	 *
+	 * This will ensure that console errors are generated, and we can surface these to the
+	 * end user in a responsible manner so that they can update their plugins and theme,
+	 * or make a decision to switch to other plugin/theme if no updates are available.
+	 */
+	public static function replace_scripts( $scripts ) {
+		$assets_url = plugins_url( 'js/', __FILE__ );
+
+		self::set_script( $scripts, 'jquery-migrate', $assets_url . 'jquery-migrate-1.4.1-wp.js', array(), '1.4.1-wp' );
+		self::set_script( $scripts, 'jquery', false, array( 'jquery-core', 'jquery-migrate' ), '1.12.4-wp' );
```

Comment:
Should the dependency not be that `jquery-migrate` needs `jquery` to be loaded first? (At least the documentation shows that as the correct load order)

Reply:
Generally yes, but there is an exception in this case; `jquery-migrate` is not a separate script and was never meant to be loaded directly. It may or may not be present/needed depending on the version of jQuery.

Also, if we add `jquery` as a dependency for `jquery-migrate`, and another plugin restores the previous code where  `jquery-migrate` is a dependency for `jquery`, it will cause a circular dependency (infinite loop). So best is to keep the way the dependencies were set in WP 5.4 and earlier.